src/services/codecovEventMetrics/useStoreCodecovEventMetric.tsx

```diff
@@ -0,0 +1,103 @@
+import { useMutation } from '@tanstack/react-query'
+import { useParams } from 'react-router-dom'
+
+import Api from 'shared/api'
+
+const mutationQuery = `
+  mutation storeEventMetric(
+    $input: StoreEventMetricsInput!
+  ) {
+    storeEventMetric(
+      input: $input
+    ) {
+      error {
+        ... on UnauthenticatedError {
+          message
+        }
+        ... on ValidationError {
+          message
+        }
+      }
+    }
+  }
+`
+
+interface Params {
+  provider: string
+}
+
+interface StoreEventMetricMutationArgs {
+  owner: string
+  event: string
+  jsonPayload: object
+}
+
+const LOCAL_STORAGE_KEY = 'UserOnboardingMetricsStored'
+const MAX_ENTRIES = 30
+
+const generateMetricString = (
+  owner: string,
+  event: string,
+  jsonPayload: string
+) => {
+  return `${owner}|${event}|${jsonPayload}`
+}
+
+const getStoredMetrics = (): string[] => {
+  const storedMetrics = localStorage.getItem(LOCAL_STORAGE_KEY)
+  return storedMetrics ? JSON.parse(storedMetrics) : []
+}
+
+const isMetricInLocalStorage = (metricString: string) => {
+  const storedMetrics = getStoredMetrics()
+  return storedMetrics.includes(metricString)
+}
+
+const addMetricToLocalStorage = (metricString: string) => {
+  let storedMetrics = getStoredMetrics()
+
+  // limit size of metrics stored to ensure we are in reasonable localstorage limits
+  if (storedMetrics.length >= MAX_ENTRIES) {
+    storedMetrics = storedMetrics.slice(1)
+  }
+
+  storedMetrics.push(metricString)
+  localStorage.setItem(LOCAL_STORAGE_KEY, JSON.stringify(storedMetrics))
+}
+
+export const useStoreCodecovEventMetric = () => {
+  const { provider } = useParams<Params>()
+
+  const mutation = useMutation({
+    mutationFn: async ({
+      owner,
+      event,
+      jsonPayload,
+    }: StoreEventMetricMutationArgs) => {
+      const jsonPayloadString = JSON.stringify(jsonPayload)
+      const metricString = generateMetricString(owner, event, jsonPayloadString)
+      if (isMetricInLocalStorage(metricString)) {
+        return
+      }
+
+      const variables = {
+        input: {
+          orgUsername: owner,
+          eventName: event,
+          jsonPayload: jsonPayloadString,
+        },
+      }
+
+      addMetricToLocalStorage(metricString)
+
+      return Api.graphqlMutation({
```

Comment:
At my last job, we rolled our own metrics for a time. One problem we ran into was metrics not getting sent for hard navigations and redirects because the redirect would happen before the request could be sent. Have you tested at all with that? Does TS query handle this?

IIRC, our solution was to use `sendBeacon`.

Reply:
I tested collection on every page and the DB updated as expected. This is good information and I will keep it in mind should something not appear to be working. 🙏

Reply:
Sounds good!